src/main/java/org/folio/circulation/resources/RequestByInstanceIdResource.java

```diff
@@ -206,16 +229,90 @@ private void createInstanceLevelRequests(RoutingContext routingContext) {
     if (items == null ||items.isEmpty()) {
       return failedValidation("Items list is null or empty", "items", "null");
     }
 
-    return succeeded(items.stream()
-      .filter(Item::isAvailable)
-      .collect(Collectors.toList()));
+    requestRelatedRecords.setUnsortedAvailableItems(
+      items.stream()
+        .filter(Item::isAvailable)
+        .collect(Collectors.toList()));
+
+    requestRelatedRecords.setUnsortedUnavailableItems(
+      items.stream()
+        .filter(item -> !item.isAvailable())
+        .collect(Collectors.toList()));
+
+    return succeeded(requestRelatedRecords);
   }
 
+  private static CompletableFuture<Result<InstanceRequestRelatedRecords>> combineWithUnavailableItems(InstanceRequestRelatedRecords records, Clients clients){
+
+    RequestQueueRepository queueRepository = RequestQueueRepository.using(clients);
+
+    final Collection<Item> unsortedUnavailableItems = records.getUnsortedUnavailableItems();
+
+    Map<Item, CompletableFuture<Result<RequestQueue>>> itemRequestQueueMap = new HashMap<>();
+    if (unsortedUnavailableItems == null || unsortedUnavailableItems.isEmpty()) {
+      return CompletableFuture.completedFuture(succeeded(records));
+    }
+
+    for (Item item : unsortedUnavailableItems) {
+      itemRequestQueueMap.put(item, queueRepository.getLiteRequestQueues(item.getItemId()));
+    }
+
+    final Collection<CompletableFuture<Result<RequestQueue>>> requestQueueFutures = itemRequestQueueMap.values();
+
+    //Collect the RequestQueue objects once they come back
+    return CompletableFuture.allOf(requestQueueFutures.toArray(new CompletableFuture[requestQueueFutures.size()]))
+      .thenApply(x -> {
+        Map<Item, RequestQueue> itemQueueSizeMap = new HashMap<>();
+        for (Map.Entry<Item, CompletableFuture<Result<RequestQueue>>> entry : itemRequestQueueMap.entrySet()) {
+          Result<RequestQueue> requestQueueResult = entry.getValue().join();
+          if (requestQueueResult.succeeded()) {
+            itemQueueSizeMap.put(entry.getKey(), requestQueueResult.value());
+          }
+        }
+
+        //Sort the map
+        Map<Item, RequestQueue> sortedMap = sortRequestQueues(itemQueueSizeMap);
+        Set<Item> sortedUnavailableItems = sortedMap.keySet();
+
+        records.setSortedUnavailableItems(new ArrayList<>(sortedUnavailableItems));
+        return succeeded(records);
+      });
+  }
+
+  private static Map<Item, RequestQueue> sortRequestQueues(Map<Item,RequestQueue> unsortedItems){
+
+    List<Map.Entry<Item, RequestQueue> > list = new LinkedList<>(unsortedItems.entrySet());
+
+    // Sort the list
+    Collections.sort(list, (q1, q2) -> {
+      RequestQueue queue1 = q1.getValue();
+      RequestQueue queue2 = q2.getValue();
+
+      int result = queue1.size() - queue2.size();
+      if (result == 0) {
+        result = queue1.getLowestPriorityFulfillableRequest()
```

Comment:
The issues references the `due date`, can you help me understand how that maps to expiration date?

Reply:
As alluded to earlier in the summary, neither approach works perfectly to predict the shortest wait time for the patron. I think requirement wise we have some wiggle room here. It's logical to compare the queue sizes, but then switching over to due date doesn't make sense to me. If a queue has 5 items but shorter due date on the current checkout item, but the other queue has 4 items and a much longer due date on its current checked out item, we'd pick the item with queue  size of 4.  Therefore I chose the last item's expiration date, but this isn't perfect. We could choose the minimum of all item's expiration date rather than the last item's. I had to make this choice because it wasn't ideal to ask Magda at 5am :) and it would require more code to get the loans up front.

Reply:
> neither approach works perfectly to predict the shortest wait time for the patron

Agreed, I'm only asking in order to understand what approach we have chosen to take.

> I think requirement wise we have some wiggle room here. It's logical to compare the queue sizes, but then switching over to due date doesn't make sense to me

Has this change in intended behaviour been discussed with Magda or Cate?

> I had to make this choice because it wasn't ideal to ask Magda at 5am :)

Ok, it sounds like it hasn't. Is she around now to ask?

If the proposed solution isn't viable, I'd prefer we just chose the first item, rather than invented an alternative approach. An improved approach can be added later.

Reply:
Yes, I discussed this with Magda afterward yesterday, and she was fine with it. She said 

> Whatever  we  pick the result will not be ideal and in many cases inaccurate because the patron can always return the item before their due date, the item with the shortest can go missing.  However from the data we found in survey conducted by RA SIG in most cases the queues are short if any. We also have other projects going in parallel that should alleviate the issue of  inaccuracy of calculating first available item.

Reply:
> We also have other projects going in parallel that should alleviate the issue of inaccuracy of calculating first available item.

Did she mention what those projects are?

Reply:
Ok, it's not my call, so I'll step away.

(Not relying on the contents of the queue could've given us options for how to do the processing to simplify the implementation)

Reply:
> (Not relying on the contents of the queue could've given us options for how to do the processing to simplify the implementation)

What other options do you have in mind?

Reply:
If we only cared about the length of the queue, we could only fetch that. Then it would be a number comparison only.

And any ties could be resolved with the first item or due date (as originally requested).

It would remove the need for the lighter way to fetch the queue, and probably some of the mapping and carrying of lists around. When we have faceting, it could likely be done with a single request.

Anyway, it isn't important any more given the interim decisions made.

Reply:
After more discussions we've decided to only care about the length of the queue because neither tie breaker approach discussed buys us much while introducing more code and adding more risks. Tie breaker is simply get the first item from the list, whichever it is.